data/transactions/logic/eval.go

```diff
@@ -4060,3 +4061,109 @@ func opBase64Decode(cx *EvalContext) {
 	}
 	cx.stack[last].Bytes, cx.err = base64Decode(cx.stack[last].Bytes, encoding)
 }
+func hasDuplicateKeys(jsonText []byte) (bool, error) {
+	dec := json.NewDecoder(bytes.NewReader(jsonText))
+	keys := make(map[string]bool)
+	var value interface{}
+	_, err := dec.Token()
+	if err != nil {
+		return false, err
+	}
+	for dec.More() {
+		// Get JSON key
+		key, err := dec.Token()
+		if err != nil {
+			return false, err
+		}
+		// end of json
+		if key == '}' {
+			break
+		}
+		// decode value
+		err = dec.Decode(&value)
+		if err != nil {
+			return false, err
+		}
+		// check for duplicates
+		if _, ok := keys[key.(string)]; ok {
+			return true, nil
+		}
+		keys[key.(string)] = true
+	}
+	return false, nil
+}
+
+func parseJSON(jsonText []byte) (map[string]json.RawMessage, error) {
+	if !json.Valid(jsonText) {
+		return nil, fmt.Errorf("invalid json text")
+	}
+	// decode text
+	decoder := json.NewDecoder(bytes.NewReader(jsonText))
+	var parsed map[string]json.RawMessage
+	err := decoder.Decode(&parsed)
+	if err != nil {
+		return nil, err
+	}
+	// check for duplicate keys
+	hasDuplicates, err := hasDuplicateKeys(jsonText)
```

Comment:
we kind of decoding json twice, `decoder.Decode(&parsed)` and inside `hasDuplicateKeys`, not good

Reply:
Unfortunately, I don't see a way to check for duplicate keys without parsing it.

Reply:
Two options: 1) Can't `parsed` map be checked? 2) use the `dec.More` + `dec.Token` for parsing and counting.

Reply:
I see. I think option 1 should work.

Reply:
actually, I went with option 2 instead.